Prototype/bt 154 treasury reward

roy-powerfinance opened the pull request:
Prototype code to award all transaction fee to all validators:
* Rewards are ```SplitTwoWays``` by 0 part to the treasury and 1 part to the validators
* Reward is then divided by the currently elected validators
* Each validator are rewarded their share with ```deposit_creating```

Note: This PR is for the team to review (optionally) and will be closed as we started to work on the actual implementation story

roy-powerfinance committed:
Partial work for rewarding all transaction fees to all validators

Commit:
Fix build issues. Fix formatting.

hoani commented on runtime/src/lib.rs:
```diff
@@ -203,11 +206,38 @@ parameter_types! {
+	fn on_nonzero_unbalanced(amount: NegativeImbalance) {
+		let validators = pallet_staking::Module::<Runtime>::current_elected();
+		if validators.len() == 0 {
+			return;
```
[thoughts for later] Might still need to handle this case, so that we don't have a permanent imbalance between total issuance and actual issuance

aliXsed replied:
Do you think this case would be likely to happen?

hoani commented on runtime/src/lib.rs:
```diff
@@ -203,11 +206,38 @@ parameter_types! {
+		if validators.len() == 0 {
+			return;
+		}
+		let per_validator_reward = amount.peek() / (validators.len() as Balance);
```
[thoughts for later] Will need to think what we want to do with the remainder so that total issuance is correct.

eg, reward = 1000, # validators = 9
reward per validator = 111, remainder = 1

aliXsed replied:
My understanding is that the inflation rate that we target at the end of an era would take care of the little discrepancies here.

mzxyz replied:
Agree with Alex, just keep it simple and the ignore the remainder will not have effect on the current mechanism

mzxyz commented on runtime/src/lib.rs:
```diff
@@ -203,11 +206,38 @@ parameter_types! {
+	<Runtime as frame_system::Trait>::AccountId,
+>>::NegativeImbalance;
+
+pub struct Validators;
```
Better to move this part of code to `impl.rs`

roy-powerfinance committed:
Moved the validator struct from libs to impl
Added logic to accumulate positive imbalance so rewards distributed to
the validators are tracked.

roy-powerfinance committed:
Merge remote-tracking branch 'origin/develop' into prototype/bt_154_treasury_reward

* origin/develop:
  Better Transaction Payment Errors (#142)
  Add setup for integration tests and 12 tests (#140)
  update return error for buy_fee_asset (#141)
  Add docker link (#136)
  Update to the changes in BuyFeeAsset (#138)

# Conflicts:
#	runtime/tests/tests.rs
#	testing/src/keyring.rs

jordy25519 commented on runtime/Cargo.toml:
```diff
@@ -47,6 +47,7 @@ frame-executive = { git = "https://github.com/plugblockchain/plug-blockchain", d
 pallet-authority-discovery = { git = "https://github.com/plugblockchain/plug-blockchain", default-features = false, branch = "1.0.0-rc2" }
 pallet-authorship = { git = "https://github.com/plugblockchain/plug-blockchain", default-features = false, branch = "1.0.0-rc2" }
 pallet-babe = { git = "https://github.com/plugblockchain/plug-blockchain", default-features = false, branch = "1.0.0-rc2" }
+pallet-balances = { git = "https://github.com/plugblockchain/plug-blockchain", default-features = false, branch = "1.0.0-rc2" }
```
Can remove this

roy-powerfinance replied:
Removed in feature/bt_154_treasury_reward branch

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,37 @@ use sp_runtime::{
+type GenericAsset<T> = pallet_generic_asset::Module<T>;
+type Staking<T> = pallet_staking::Module<T>;
+
+pub struct Validators;
```
Could use a more descriptive name + comment

roy-powerfinance replied:
Renamed to SplitToAllValidators and added comments
 in feature/bt_154_treasury_reward branch

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,37 @@ use sp_runtime::{
+	fn on_nonzero_unbalanced(amount: NegativeImbalance) {
+		let validators = Staking::<Runtime>::current_elected();
+		if validators.len() > 0 {
+			let per_validator_reward:Balance = amount.peek() / (validators.len() as Balance);
```
`per_validator_reward: Balance`

roy-powerfinance replied:
i will run cargo fmt

roy-powerfinance replied:
Changed in the new feature branch

jordy25519 commented on runtime/src/lib.rs:
```diff
@@ -203,11 +206,22 @@ parameter_types! {
 	pub const WeightFeeCoefficient: Balance = 1_000;
 }
 
+type NegativeImbalance = <<Runtime as crml_transaction_payment::Trait>::Currency as Currency<
```
Re-use the version from `impls.rs` or vice versa

roy-powerfinance replied:
libs now references negativeImbalance defined in impl
Added this to the new feature branch

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,37 @@ use sp_runtime::{
+			for v in &validators {
+				let payout = <GenericAsset<Runtime> as MultiCurrencyAccounting>::deposit_creating(
+					&v,
+					Some(amount.asset_id().into()),
```
I think we should unroll this by getting the `asset_id` outside the loop once

roy-powerfinance replied:
Changed this in feature/bt_154_treasury_reward branch

hoani commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,37 @@ use sp_runtime::{
+			}
+
+		}
+	}
```
Suggest keeping track of any unrewarded imbalance and burning it at the end of this function. 

This will also take care case where validators.len() == 0.

hoani replied:
Spoke with Roy, this is handled

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,37 @@ use sp_runtime::{
+					&v,
+					Some(amount.asset_id().into()),
+					per_validator_reward);
+				total_imbalance.subsume(payout);
```
👍

jordy25519 commented on runtime/src/impls.rs:
```diff
@@ -35,6 +36,37 @@ use sp_runtime::{
+impl OnUnbalanced<NegativeImbalance> for Validators
+{
+	fn on_nonzero_unbalanced(amount: NegativeImbalance) {
+		let validators = Staking::<Runtime>::current_elected();
```
if block fees are 0 we could also return

roy-powerfinance replied:
Added in feature/bt_154_treasury_reward branch

jordy25519 reviewed (commented):
Looks good overall! Some minor nitpicks

roy-powerfinance commented:
All further work is done in feature/bt_154_treasury_reward branch.

mzxyz commented:
This PR can be closed
